### PR TITLE
miniupnpd: bump version, drop igdv1 variant, add nftables support

### PR DESCRIPTION
**Pull request by stintel**
Maintainer: NA
Compile/run tested: OpenWrt master r18045-71af55ae2a on [qoriq](https://git.openwrt.org/?p=openwrt/staging/stintel.git;a=shortlog;h=refs/heads/qoriq)

Please do not merge this before someone tested the iptables variant. I have currently only tested the nftables variant.

### CODE REVIEW

**Comment by stintel** [comment]
Please do not merge this before someone tested the iptables variant. I have currently only tested the nftables variant.

**Comment by dangowrt** [comment]
`libnetfilter-conntrack3` still pulls in `kmod-ipt-conntrack`... Apparently `libnetfilter-conntrack3` isn't even needed for the nftables variant:
```
root@bpi-r64:/# ldd /usr/sbin/miniupnpd
	/lib/ld-musl-aarch64.so.1 (0x7f802fe000)
	libcap-ng.so.0 => /usr/lib/libcap-ng.so.0 (0x7f802ad000)
	libnftnl.so.11 => /usr/lib/libnftnl.so.11 (0x7f8026b000)
	libmnl.so.0 => /usr/lib/libmnl.so.0 (0x7f80255000)
	libuuid.so.1 => /usr/lib/libuuid.so.1 (0x7f8023e000)
	libc.so => /lib/ld-musl-aarch64.so.1 (0x7f802fe000)
	libgcc_s.so.1 => /lib/libgcc_s.so.1 (0x7f8021b000)
```

**Comment by stintel** [comment]
> `libnetfilter-conntrack3` still pulls in `kmod-ipt-conntrack`... Apparently `libnetfilter-conntrack3` isn't even needed for the nftables variant:

Thanks. I've incorrectly moved back the dependencies due to miniupnpd hardcoding iptables when detecting OpenWrt. Didn't check the dependencies anymore after fixing that.

File: net/miniupnpd/Makefile [hunk 1]
```diff
@@ -26,27 +26,47 @@ PKG_BUILD_PARALLEL:=1
...
+	+libnetfilter-conntrack
+  TITLE+= (iptables)
+  CONFLICTS:=miniupnpd-nftables
+  VARIANT:=iptables
+endef
+
+define Package/miniupnpd-nftables
+  $(call Package/miniupnpd/Default)
+  DEPENDS+= \
+	+nftables
```
Comment (dangowrt) [inline net/miniupnpd/Makefile:57]:
```suggestion
+	+libnftnl
```

File: net/miniupnpd/Makefile [hunk 2]
```diff
@@ -9,7 +9,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=miniupnpd
 PKG_VERSION:=2.2.3
-PKG_RELEASE:=1
+PKG_RELEASE:=2
```
Comment (dangowrt) [inline net/miniupnpd/Makefile:13]:
```suggestion
+PKG_USE_MIPS16:=0

```

**Comment by stintel** [comment]
Temporarily removed all but the version bump commit from this PR to test if that's what's breaking MIPS16.

File: net/miniupnpd/Makefile (hunk 2 above)
Comment (stintel) [reply net/miniupnpd/Makefile:13]: PKG_USE_MIPS16:=0 doesn't seem to help :(

Comment (stintel) [reply net/miniupnpd/Makefile:13]: Disabling LTO for the nftables variant did.

**Comment by dangowrt** [comment]
Tested the nftables variant and while it does seem to work and also create appropriate forwarding rules in it's tables, there is still something missing because the rules there don't have any effect.
```
root@OpenWrt:/# nft list tables
table inet fw4
table inet miniupnpd
table ip miniupnpd
table ip6 miniupnpd
root@OpenWrt:/# nft list table inet miniupnpd
table inet miniupnpd {
	chain forward {
		type filter hook forward priority -25; policy accept;
		iif "wwan0" th dport 1234 @nh,128,32 3232236975 @nh,72,8 6 accept
		iif "wwan0" th dport 1234 @nh,128,32 3232236975 @nh,72,8 17 accept
	}
}
root@OpenWrt:/# nft list table ip miniupnpd
table ip miniupnpd {
	chain prerouting {
		type nat hook prerouting priority dstnat; policy accept;
		iif "wwan0" tcp dport 1234 dnat to 192.168.5.175:1234
		iif "wwan0" udp dport 1234 dnat to 192.168.5.175:1234
	}

	chain postrouting {
		type nat hook postrouting priority srcnat; policy accept;
	}
}
```

I assume we still need a hook in fw4 to jump to the miniupnpd table. [This](https://github.com/miniupnp/miniupnp/blob/master/miniupnpd/netfilter_nft/scripts/nft_init.sh) can provide some hints maybe...

**Comment by stintel** [comment]
> Tested the nftables variant and while it does seem to work and also create appropriate forwarding rules in it's tables, there is still something missing because the rules there don't have any effect.

The added rules work fine for me.

**Comment by dangowrt** [comment]
> The added rules work fine for me.

With `fw4` running at the same time?

**Comment by stintel** [comment]
> With `fw4` running at the same time?

Yes. My main router is fully migrated to fw4. My Xbox (the reason for using miniupnpd) has "open NAT".

**Comment by stintel** [comment]
What does seem to be a problem is that these miniupnpd rules get duplicated every time a device requests a upnp mapping:

```
table inet miniupnpd {
        chain forward {
                type filter hook forward priority -25; policy accept;
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
                iif "wan" th dport 3074 @nh,128,32 3232249379 @nh,72,8 17 accept
        }
}
table ip miniupnpd {
        chain prerouting {
                type nat hook prerouting priority dstnat; policy accept;
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
                iif "wan" udp dport 3074 dnat to 192.168.54.35:3074
        }

        chain postrouting {
                type nat hook postrouting priority srcnat; policy accept;
        }
}
```

**Comment by aparcar** [comment]
Is this work in progress? If so please mark it as a draft.

**Comment by stintel** [comment]
I am not considering this a work in progress. Yes, the nftables variant might have some bugs, but the iptables variant should not change behaviour. I am using the nftables variant on my main router for more than a month and it is usable. I created an upstream issue, these fixes can be added later imo.

**Comment by stintel** [comment]
Fixed a missing PROVIDES for the iptables variant and did basic runtime test on my backup router, which is still running firewall3. If no objections, I will merge this soon.

**Comment by stintel** [comment]
Dropped the bump as apparently that was merged from a later PR.

**Comment by theAeon** [comment]
Thoughts on backporting the version bump to 21.02 for the igd fix?

**Comment by neilsan1366** [comment]
is what this problem is solved ? i'm use a belkin rt3200 with upnp and firewall 4, thanks

**Comment by hnyman** [comment]
@stintel @dangowrt @aparcar 

Now that firewall4 with nftables is the default in master, should we declare the nftables variant as the DEFAULT_VARIANT for miniupnpd?   

Otherwise opkg may find the iptables variant first and install it. 
(The iptables variant may also be acidentally selected by dependent packages, e.g. the LuCI app.)

**Comment by aparcar** [comment]
Sounds good to me.

**Comment by coudu** [comment]
> Tested the nftables variant and while it does seem to work and also create appropriate forwarding rules in it's tables, there is still something missing because the rules there don't have any effect.

same for me, rules are there but have no effect...
Also the init script has some iptables commands, wouldn't the nftables variant need a different init script? or does it need iptables-nft?

**Comment by hnyman** [comment]
The Nftables dependencies are still screwed to some extent.
See pr https://github.com/openwrt/openwrt/pull/5004 and https://github.com/openwrt/openwrt/pull/5007
Iptables-nft will probably be a seamless drop-in replacement in a few days/weeks, once those PRs from @champtar  and @tiagogaspar8  are merged.

**Pull request merged**